### PR TITLE
Extended border-radius functionality - Issue #2602

### PR DESCRIPTION
**Pull request by bansalnitish**
Fixes - mozilla/thimble.mozilla.org#2602
Quick Edit UI for the following is implemented in this PR :
 
- border-top-left-radius
- border-top-right-radius
- border-bottom-left-radius
- border-bottom-right-radius

### CODE REVIEW

File: src/extensions/default/InlineBorderRadiusEditor/BorderRadiusEditor.js [hunk 1]
```diff
@@ -66,9 +66,14 @@ define(function(require, exports, module) {
         return this.value + (this.value === 0 ? "" : this.unit);
     };
 
-    function BorderRadiusEditor($parent, valueString, radiusChangeHandler) {
+    function BorderRadiusEditor($parent, valueString, radiusChangeHandler, type) {
         var self = this;
-
+        if(type === "border-radius"){
+            BorderRadiusTemplate = require("text!BorderRadiusEditorTemplate.html");
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/BorderRadiusEditor.js]: the templates are basically the same so let's not create two of them.

Instead, try this pattern:
1. Add a function to `BorderRadiusUtils` called `getSingleProperty`. The function will look something like:
```js
function getSingleProperty(property) {
    var property = /top-left|top-right|bottom-left|bottom-right/.exec(property);
    return property && property[0];
}
```
2. Let this function take two extra arguments - one called `isSingleProperty` that is a boolean and another called `iconClassName`. In this function let the default value for the first be `false`. Before calling this function, get the value of `BorderRadiusUtils.getSingleProperty` by passing the css property to it. Store it in a variable called `singleProperty`. Pass the value of `!!singleProperty` as the `isSingleProperty` argument to this function, and pass the value of `singleProperty` as the `iconClassName` argument.
3. In general the idea is to just use the all-corners logic for the single values as well and disable the individual value logic. First, change the template to conditionally render certain parts of the template. To do that, wrap each part that needs to be conditionally rendered in a [mustache section](https://github.com/janl/mustache.js#sections). For e.g., do this in the template
```html
{{#individualCorners}}
<div id="individualCornerArea">
    ...
</div>
{{/individualCorners}}
```
and
```html
{{#inlineEditorHeader}}
<header>
    <ul class="button-bar">
        ...
    </ul>
</header>
{{/inlineEditorHeader}}
```
and
```html
{{#allCornerIcon}}
<div class="all corner-icon">
    <div class="dot"></div>
    <div class="dot"></div>
    <div class="dot"></div>
    <div class="dot"></div>
</div>
{{/allCornerIcon}}
{{#singleCornerIcon}}
<div class="{{iconClassName}} corner-icon"></div>
{{/singleCornerIcon}}
```
4. Then in this function, you modify the value of strings to determine what to render. Your code will look something like this:
```js
var values = JSON.parse(JSON.stringify(Strings));

if (!isSingleProperty) {
    values["inlineEditorHeader"] = true;
    values["individualCorners"] = true;
    values["allCornerIcon"] = true;
} else {
    values["singleCornerIcon"] = true;
}

self.$element = $(Mustache.render(BorderRadiusTemplate, values));
```
The rest of the logic should handle itself the way it is. Try it out and feel free to ask questions.

Comment (bansalnitish) [reply src/extensions/default/InlineBorderRadiusEditor/BorderRadiusEditor.js]: Thank you so much @gideonthomas for such awesome guidance. I'll try this one out.

**Comment by bansalnitish** [comment]
@gideonthomas thank you for suggesting those cool tricks. I have made the changes. Can you just give a pass on the code !

**Comment by bansalnitish** [comment]
@gideonthomas any updates on this ?

**Comment by bansalnitish** [comment]
@gideonthomas Can you please review this one ?

File: src/extensions/default/InlineBorderRadiusEditor/BorderRadiusEditorTemplate.html [hunk 2]
```diff
@@ -117,5 +122,32 @@
                </td>
            </tr>
        </table>
-    <div>
+    </div>
+ {{/allCornerArea}}  
+
+ {{#singleCornerArea}}
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/BorderRadiusEditorTemplate.html]: let's not do this! Almost all of the code in this section is the same as the `allCornerArea` section. So, just use the `allCornerArea` section and conditionally render the icon (which is the only difference). See the last part of the third point in my comment here: https://github.com/mozilla/brackets/pull/920#discussion_r164550664

File: src/extensions/default/InlineBorderRadiusEditor/BorderRadiusUtils.js [hunk 3]
```diff
@@ -49,9 +49,15 @@ define(function (require, exports, module) {
     // Matches a single value and captures the number and unit. Use it with exec()
     // to find successive values in a valid border radius value string.
     var BORDER_RADIUS_SINGLE_VALUE_REGEX = new RegExp(/(\d+\.?\d*)(px|em|%)?/, "g");
+    // Get sinlge property if it is for the border radius
+    function getSingleProperty(property) {
+    var property = /top-left|top-right|bottom-left|bottom-right/.exec(property);
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/BorderRadiusUtils.js]: fix your indentation here

File: src/extensions/default/InlineBorderRadiusEditor/main.js [hunk 4]
```diff
@@ -66,11 +67,14 @@ define(function (require, exports, module) {
 
             // Get the css property name after removing spaces and ":" so that we can check for it in the file BorderRadiusProperties.json
             cssPropertyName = cursorLine.split(':')[0].trim();
-
+
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/main.js]: undo this

File: src/extensions/default/InlineBorderRadiusEditor/main.js [hunk 5]
```diff
@@ -31,7 +31,8 @@ define(function (require, exports, module) {
         BorderRadiusUtils = require("BorderRadiusUtils");
 
     var DEFAULT_RADIUS = "15px";
-
+    var isSingleProperty = false;
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/main.js]: this should not be a global

File: src/extensions/default/InlineBorderRadiusEditor/css/main.less [hunk 6]
```diff
@@ -15,6 +15,12 @@
         padding-bottom: 10px;
         font-size: 14px;
     }
+    label {
+            display: inline-block;
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/css/main.less]: fix your indentation

File: src/extensions/default/InlineBorderRadiusEditor/main.js [hunk 7]
```diff
@@ -31,7 +31,8 @@ define(function (require, exports, module) {
...
+    var singleProperty;
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/main.js]: what's this? this shouldn't be a global

File: src/extensions/default/InlineBorderRadiusEditor/BorderRadiusUtils.js [hunk 8]
```diff
@@ -49,9 +49,15 @@ define(function (require, exports, module) {
     // Matches a single value and captures the number and unit. Use it with exec()
     // to find successive values in a valid border radius value string.
     var BORDER_RADIUS_SINGLE_VALUE_REGEX = new RegExp(/(\d+\.?\d*)(px|em|%)?/, "g");
+    // Get sinlge property if it is for the border radius
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/BorderRadiusUtils.js]: single*

File: src/extensions/default/InlineBorderRadiusEditor/main.js [hunk 9]
```diff
@@ -66,11 +67,14 @@ define(function (require, exports, module) {
 
             // Get the css property name after removing spaces and ":" so that we can check for it in the file BorderRadiusProperties.json
             cssPropertyName = cursorLine.split(':')[0].trim();
-
+            
             if (!cssPropertyName || !properties[cssPropertyName]) {
                 return null;
             }
 
+            singleProperty = BorderRadiusUtils.getSingleProperty(cssPropertyName);
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/main.js]: rename this to `cssClassSuffix`.

File: src/extensions/default/InlineBorderRadiusEditor/main.js [hunk 10]
```diff
@@ -66,11 +67,14 @@ define(function (require, exports, module) {
...
+            isSingleProperty = !!singleProperty;
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/main.js]: you don't need globals for this. Just declare them here and return them in the object returned. Then, in the calling function, use `context.cssClassSuffix`, etc. to access these values.

File: src/extensions/default/InlineBorderRadiusEditor/InlineBorderRadiusEditor.js [hunk 11]
```diff
@@ -40,12 +40,14 @@ define(function (require, exports, module) {
      * @param {!string} borderRadius  Initially selected borderRadius
      * @param {!CodeMirror.TextMarker} marker
      */
-    function InlineBorderRadiusEditor(borderRadius, marker) {
+    function InlineBorderRadiusEditor(borderRadius, marker, singleProperty, isSingleProperty) {
```
Comment (gideonthomas) [inline src/extensions/default/InlineBorderRadiusEditor/InlineBorderRadiusEditor.js]: rename `singleProperty` to `cssClassSuffix`

File: src/extensions/default/InlineBorderRadiusEditor/BorderRadiusUtils.js (hunk 8 above)
Comment (bansalnitish) [reply src/extensions/default/InlineBorderRadiusEditor/BorderRadiusUtils.js]: oh my bad !

**Comment by bansalnitish** [comment]
Hey @gideonthomas, I have made the requested changes.

**Comment by humphd** [comment]
@gideonthomas are we good to go on this?

**Comment by bansalnitish** [comment]
@gideonthomas Can you have a look on this PR ?

**Comment by gideonthomas** [comment]
apologies for the delay again! Yes this looks great! Thanks for the contribution @bansalnitish

**Comment by bansalnitish** [comment]
No Problem @gideonthomas 
Thank you for merging this one !
Cheers

**Pull request merged**